shotover/src/server.rs

```diff
@@ -255,6 +263,138 @@ pub struct Handler<C: CodecBuilder> {
+    codec: C,
+    ws_stream: WebSocketStream<TcpStream>,
+    in_tx: UnboundedSender<Messages>,
+    mut out_rx: UnboundedReceiver<Messages>,
+    out_tx: UnboundedSender<Messages>,
+) {
+    let (mut writer, mut reader) = ws_stream.split();
+    let (mut decoder, mut encoder) = codec.build();
+
+    // read task
+    tokio::spawn(async move {
+        loop {
+            tokio::select! {
+                result = reader.next() => {
+                    if let Some(ws_message) = result {
+                        match ws_message {
+                            Ok(WsMessage::Binary(ws_message_data)) => {
+                                // Entire message is reallocated and copied here due to incompatibility between tokio codecs and tungstenite.
+                                let message = decoder.decode(&mut BytesMut::from(ws_message_data.as_slice()));
+                                match message {
+                                    Ok(Some(message)) => {
+                                        if in_tx.send(message).is_err() {
+                                            // main task has shutdown, this task is no longer needed
+                                            return;
+                                        }
+                                    }
+                                    Ok(None) => {
+                                        // websocket client has closed the connection
+                                        return;
+                                    }
+                                    Err(CodecReadError::RespondAndThenCloseConnection(messages)) => {
+                                        if let Err(err) = out_tx.send(messages) { // TODO we need
+                                                                                  // to send a
+                                                                                  // close message
+                                                                                  // to the client
+                                            error!("Failed to send RespondAndThenCloseConnection message: {:?}", err);
+                                        }
+                                        return;
+                                    }
+                                    Err(CodecReadError::Parser(err)) => { // TODO we need to send a
+                                                                          // close message to the
+                                                                          // client, protocol error
+                                        warn!("failed to decode message: {:?}", err);
+                                        return;
+                                    }
+                                    Err(CodecReadError::Io(_err)) => {
+                                        unreachable!("CodecReadError::Io should not occur because we are reading from a newly created BytesMut")
```

Comment:
I dont understand this unreachable.
What is the BytesMut it is referring to?
An IO error sounds very reachable to me.

Reply:
line 285, we put the read data from the websocket binary message into a BytesMut.

Reply:
oooh I see, the actual IO error would happen on line 321 :+1: